README.md
# Timer

## 计时器

[API documentation](https://lisfan.github.io/timer/)

## Feature 特性

- 支持不同两种计时模式：递增计时和递减计时（倒计时），默认为倒计时
- 可检测到时间流速不正常，如待机、休眠、或者切换了浏览器标签，造成进程挂起，计时不准确时，自动修正时间流速错误，调整回正确的时间

## Detail 详情

- 计时时间戳选项接收值单位为毫秒，且是除以1000弃余后的整数，即超出时会向下取整
- 提供妙表记录功能
- 暂不支持毫秒计时

## Install 安装

```bash
npm install -S @~lisfan/timer
```

## Usage 起步

```js
import Timer from '@~lisfan/timer'

const timer = new Timer({
   timeStamp: 1 * 10 * 1000, // 创建一个10秒的计时器
})

// 开始计时
timer.start((timer) => {
    console.log('每秒执行一次该回调',timer.$datetime)
}).then(() => {
    console.log('计时结束')
}).catch(() => {
    console.log('中途造成计时中断')
})

// 计时暂停并开始
timer.stop()
// 计时继续
timer.start().then(...).catch(...)

// 计时复位并开始
timer.reset()
// 计时开始
timer.start().then(...).catch(...)

// 计时当前值记录
timer.record()

```

CHANGELOG.md
# CHANGELOG 更新日志

## v1.1.0-rc
### 2017.10.19
- [refactor] 小重构

## v1.0.0
### 2017.05.29
- [feature] 增加递增计时功能
- [fix] 修复计时器开始时，回调延迟一秒触发的bug


src/index.js
/**
 * @file 计时器
 * @author lisfan <goolisfan@gmail.com>
 * @version 1.1.0
 * @licence MIT
 */

import validation from '@~lisfan/validation'
import Logger from '@~lisfan/logger'

import FormatDate from './utils/format-date'
import MODE_TYPE from './enums/mode-type'

/**
 * 私有行为
 * @private
 */
const _actions = {
  /**
   * 返回向下取整后的时间戳
   * @param {number} [timeStamp=Date.now()] - 时间戳
   * @returns {number}
   */
  timeStamp(timeStamp = Date.now()) {
    return Math.floor(timeStamp / 1000) * 1000
  },
  /**
   * 格式化时间，返回formatDate实例
   * @param {Timer} self - timer实例
   * @returns {FormatDate}
   */
  formatDate(self) {
    let date

    if (self.$mode === '+') {
      date = (self.$options.timeStamp - self.$remainTimeStamp) + self.$timeZoneTimeStamp
    } else {
      date = self.$remainTimeStamp + self.$timeZoneTimeStamp
    }

    return new FormatDate({
      date,
      format: this.$format
    })
  }
}

/**
 * @classdesc
 * 计时类
 *
 * @class
 */
class Timer {
  /**
   * 默认配置选项
   *
   * @since 1.0.0
   * @static
   * @readonly
   * @memberOf Timer
   * @property {boolean} debug=false - 调试模式
   * @property {string} format='mm:ss' - 日期时间格式化字符串
   * @property {string} mode=- - 计时模式类型，可选值请参考 {@link MODE_TYPE}
   */
  static options = {
    // timeStamp: undefined, // 必须
    debug: false, // 开启调试模式
    format: 'mm:ss', // 日期时间格式化字符串
    mode: '-', // 计时模式
  }

  /**
   * 更新默认配置选项
   *
   * @since 1.0.0
   * @static
   * @param {object} options - 配置选项
   * @param {boolean} [options.debug] - 调试模式
   * @param {string} [options.format] - 日期时间格式化字符串
   * @param {string} [options.mode] - 计时模式类型，可选值请参考 {@link MODE_TYPE}
   */
  static config(options) {
    const ctor = this

    ctor.options = {
      ...ctor.options,
      ...options
    }
  }

  /**
   * 构造函数
   *
   * @param {object} options - 配置选项
   * @param {number} options.timeStamp - 剩余计时时间戳，毫秒单位，且毫秒数必须为1000单位，不能是1234这样，超出时会自动向下取整
   * @param {boolean} [options.debug] - 调试模式
   * @param {string} [options.format] -
   *   日期时间格式化字符串，支持使用字母占位符匹配对应的年月日时分秒：Y=年、M=月、D=日、h=时、m=分、s=秒、ms=毫秒，年和毫秒字母占位符可以使用1-4个，其他占位符可以使用1-2个，如果实际结果值长度大于占位符的长度，则显示值实际结果值，如果小于，则前置用0补足
   * @param {string} [options.mode] - 计时模式类型，可选值请参考 {@link MODE_TYPE}
   */
  constructor(options) {
    const ctor = this.constructor

    this.$options = {
      ...ctor.options,
      ...options
    }

    this._logger = new Logger({
      name: 'timer',
      debug: this.$options.debug
    })

    if (!this.$options.timeStamp) {
      this._logger.error('require timeStamp option param, please check!')
    }

    this.$status = 'prepare'

    // 当前时间戳
    this.$remainTimeStamp = _actions.timeStamp(this.$options.timeStamp)

    // 当前倒计时过去的时间戳
    this.$throughTimeStamp = 0

    // 当前倒计时间字符串
    // 如果是倒计时，则将当前的时间设置为最大值
    this._formatDate = _actions.formatDate(this)
  }

  /**
   * 日志打印器，方便调试
   *
   * @since 1.0.0
   * @private
   */
  _logger = undefined

  /**
   * 日期时间格式化实例
   *
   * @since 1.0.0
   * @private
   */
  _formatDate = undefined

  /**
   * 延迟计时器ID
   *
   * @since 1.0.0
   * @readonly
   */
  _timeouter = undefined

  /**
   * 实例配置项
   *
   * @since 1.0.0
   * @readonly
   */
  $options = undefined

  /**
   * 实例状态，可能处于以下状态：prepare=准备创段、processing=进行中、finished=已完成、stoped=暂停中
   *
   * @since 1.0.0
   * @readonly
   */
  $status = undefined

  /**
   * 当前地区的偏移时区时间戳
   *
   * @since 1.0.0
   * @readonly
   */
  $timeZoneTimeStamp = new Date().getTimezoneOffset() * 1000 * 60
  /**
   * 剩余计时时间戳
   *
   * @since 1.0.0
   * @readonly
   */
  $remainTimeStamp = undefined
  /**
   * 已经过的时间戳
   *
   * @since 1.0.0
   * @readonly
   */
  $throughTimeStamp = undefined

  /**
   * 开始计时时间戳
   *
   * @since 1.0.0
   * @readonly
   */
  $startTimeStamp = undefined

  /**
   * 暂停计时时间戳
   *
   * @since 1.0.0
   * @readonly
   */
  $stopTimeStamp = undefined

  /**
   * 计时器开始计时起 => 计时器完成时的一轮周期计时时间戳
   *
   * @since 1.0.0
   * @readonly
   */
  $endTimeStamp = undefined
  /**
   * 当前时间的时间戳，用于修正计时器在空间维度中出现异常流逝的时间戳
   *
   * @since 1.0.0
   * @readonly
   */
  $currentTimeStamp = undefined

  /**
   * 计时器每次记录的妙表功能
   *
   * @since 1.0.0
   * @readonly
   */
  $stopwatch = []

  /**
   * 获取实例的调试配置项
   *
   * @since 1.0.0
   * @getter
   * @returns {string}
   */
  get $debug() {
    return this._logger.$debug
  }

  /**
   * 获取实例的计时时间戳配置项
   *
   * @since 1.0.0
   * @getter
   * @readonly
   * @returns {string}
   */
  get $timeStamp() {
    return this.$options.timeStamp
  }

  /**
   * 获取实例的日期时间格式化配置项
   *
   * @since 1.0.0
   * @getter
   * @readonly
   * @returns {string}
   */
  get $format() {
    return this.$options.format
  }

  /**
   * 获取实例的计时模式配置项
   *
   * @since 1.0.0
   * @getter
   * @readonly
   * @returns {string}
   */
  get $mode() {
    return MODE_TYPE[this.$options.mode]
  }

  /**
   * 获取实例的日期时间字符串值
   *
   * @since 1.0.0
   * @getter
   * @readonly
   * @returns {string}
   */
  get $datetime() {
    return this._formatDate.toString()
  }

  /**
   * 获取实例的日期时间数据片断
   *
   * @since 1.1.0
   * @getter
   * @readonly
   * @property {string} [year] - 剩余年数
   * @property {string} [month] - 剩余月数
   * @property {string} [date] - 剩余日数
   * @property {string} [hour] - 剩余时数
   * @property {string} [minute] - 剩余分数
   * @property {string} [second] - 剩余秒数
   * @property {string} [millisecond] - 剩余毫秒数
   */
  get $data() {
    return this._formatDate.$data
  }

  /**
   * 计时器开始计时
   * 计时器结束后进入resolved状态
   * 若在未结束前中途造成暂停，会触发rejectd状态
   *
   * @since 1.0.0
   * @param {function} callback - 每秒执行的回调函数
   * @returns {Promise}
   */
  start(callback) {
    return new Promise((resolve, reject) => {
      this.$status = 'processing'

      // 记录启动时间时间戳
      this.$startTimeStamp = _actions.timeStamp()

      // 设置结束时间戳
      this.$endTimeStamp = this.$startTimeStamp + this.$remainTimeStamp

      // 计时函数
      const dingdong = () => {
        // 如果当前已暂停，则停止倒计时
        if (this.$status !== 'processing') {
          return reject(this.$status)
        }

        // 当前的超时时间戳（当currentTimeStamp不在startTimeStamp和endTmeStamp内时，说明已超时）
        // 倒计时也可能在一些途中超时挂起，导致倒计时没有再走下去，所以每次需要重新捕获
        this.$currentTimeStamp = _actions.timeStamp()

        this.$remainTimeStamp -= 1000
        this.$throughTimeStamp += 1000

        this._logger.log('throughTimeStamp', ((this.$throughTimeStamp / 1000) ) + 's')

        if (this.$currentTimeStamp > (this.$startTimeStamp + this.$throughTimeStamp)) {
          this._logger.log('超过时间流速，自动修正!')
          // 超过的时间流速
          const lostTime = this.$currentTimeStamp - this.$startTimeStamp
          this.$remainTimeStamp = this.$options.timeStamp - lostTime - 1000
          this.$throughTimeStamp = lostTime + 1000
        }

        // 计时时间已到达
        if (this.$remainTimeStamp < 0) {
          this.$remainTimeStamp = 0
          this.$throughTimeStamp = this.$endTimeStamp - this.$startTimeStamp
        }

        this._formatDate = _actions.formatDate(this)

        // 优先执行回调
        if (validation.isFunction(callback)) {
          callback.call(this)
        }

        /* eslint-enable max-len*/
        if (this._timeouter && this.$currentTimeStamp >= this.$endTimeStamp) {
          clearTimeout(this._timeouter)
          this._timeouter = undefined

          this.$status = 'finished'
          resolve(this.$status)
        } else {
          this._timeouter = setTimeout(() => {
            dingdong()
          }, 1000)
        }
      }

      this._timeouter = setTimeout(() => {
        dingdong()
      }, 1000)
    })
  }

  /**
   * 秒表：记录触发该操作的计时结果，保存结果的值将根据当前实例的format格式化
   *
   * @since 1.1.0
   * @returns {Timer}
   */
  record() {
    this.$stopwatch.push(this.$datetime)
    return this
  }

  /**
   * 计时器停止
   *
   * @since 1.0.0
   * @returns {Timer}
   */
  stop() {
    clearTimeout(this._timeouter)
    this._timeouter = undefined

    this.$status = 'stoped'
    this.$stopTimeStamp = _actions.timeStamp()

    return this
  }

  /**
   * 计时器复位
   *
   * @since 1.0.0
   * @returns {Timer}
   */
  reset() {
    clearTimeout(this._timeouter)
    this._timeouter = undefined

    // 秒表复位
    this.$stopwatch = []

    this.$status = 'prepare'

    this.$remainTimeStamp = _actions.timeStamp(this.$options.timeStamp)
    this.$startTimeStamp = undefined
    this.$stopTimeStamp = undefined
    this.$endTimeStamp = undefined
    this.$currentTimeStamp = undefined

    this.$throughTimeStamp = 0

    this._formatDate = _actions.formatDate(this)

    return this
  }
}

export default Timer

src/utils/format-date.js
/**
 * @file 计时器日期时间格式化工具
 * @author lisfan <goolisfan@gmail.com>
 * @version 1.0.0
 * @licence MIT
 * @ignore
 */

import validation from '@~lisfan/validation'
import Logger from '@~lisfan/logger'

const logger = new Logger('format-date')

// 日期时间格式化模式匹配数据片段映射
const DATETIME_PATTERN = {
  'Y': 'year',
  'M': 'month',
  'D': 'date',
  'h': 'hour',
  'm': 'minute',
  's': 'second',
  'S': 'millisecond',
}

/**
 * 计时器格式化类
 *
 * @class
 * @ignore
 */
class FormatDate {
  /**
   * 默认配置选项
   *
   * @memberOf FormatDate
   * @since 1.0.0
   * @readonly
   * @static
   */
  static options = {
    format: 'mm:ss'
  }

  /**
   * 构造函数
   *
   * @param {object} options - 配置对象
   * @param {number|string|Date} options.date - 可以被格式为时间的值
   * @param {string} [options.format='mm:ss'] -
   *   日期时间格式化字符串，支持使用字母占位符匹配对应的年月日时分秒：Y=年、M=月、D=日、h=时、m=分、s=秒、ms=毫秒，年和毫秒字母占位符可以使用1-4个，其他占位符可以使用1-2个，如果实际结果值长度大于占位符的长度，则显示值实际结果值，如果小于，则前置用0补足
   */
  constructor(options) {
    if (!validation.isNumber(options.date) && !validation.isString(options.date) && !validation.isDate(options.date)) {
      logger.error('require date option param, please check')
    }

    const ctor = this.constructor

    this.$options = {
      ...ctor.options,
      ...options
    }
  }

  /**
   * 获取日期时间格式化后的日期时间片断数据
   *
   * @since 1.0.0
   * @function
   * @param {number|string|Date} date - 可以被格式为时间的值
   * @param {string} [format='mm:ss'] - 日期时间格式化字符串
   * @returns {object}
   */
  static getFields(date, format = FormatDate.options.format) {
    if (!validation.isNumber(date) && !validation.isString(date) && !validation.isDate(date)) {
      logger.error('require date option param, please check')
    }

    const endDate = new Date(date) // 目标结束时间
    const startDate = new Date(0) // 计时器用于减去初始日期时间

    const fields = {
      'Y': endDate.getFullYear() - startDate.getFullYear(),
      'M': endDate.getMonth() - startDate.getMonth(),
      'D': endDate.getDate() - startDate.getDate(),
      'h': endDate.getHours(),
      'm': endDate.getMinutes(),
      's': endDate.getSeconds(),
      'S': endDate.getMilliseconds(),
    }

    let formatFields = {}

    Object.entries(DATETIME_PATTERN).forEach(([pattern, dateStr]) => {
      const regexp = new RegExp(pattern + '+')
      const matched = format.match(regexp)

      if (matched) {
        const value = fields[pattern].toString()
        const valueLen = value.length
        const patternLen = matched[0].length
        // 如果指定长度小于要值长度，则显示值长度
        // 如果不足，前置用0补足
        if (patternLen >= valueLen) {
          formatFields[dateStr] = value.padStart(patternLen, '0')
        } else {
          formatFields[dateStr] = value
        }
      }
    })

    return formatFields
  }

  /**
   * 获取日期时间格式化后的字符串
   *
   * @since 1.0.0
   * @param {number|string|Date|object} dateOrFields - 可以被格式为时间的值，或者已经被提前的日期时间值片断
   * @param {string} [format='mm:ss'] - 日期时间格式化字符串
   * @returns {string}
   */
  static toString(dateOrFields, format = FormatDate.options.format) {
    const ctor = this

    let data

    if (validation.isPlainObject(dateOrFields)) {
      data = dateOrFields
    } else {
      data = ctor.getFields(dateOrFields, format)
    }

    Object.entries(DATETIME_PATTERN).forEach(([pattern, dateStr]) => {
      const regexp = new RegExp((pattern + '+'))
      format = format.replace(regexp, data[dateStr])
    })

    return format
  }

  /**
   * 实例的配置项
   *
   * @since 1.0.0
   * @readonly
   * @property {number|string|Date} date - 可以被格式为时间的值
   * @property {string} format='mm:ss' - 日期时间格式化字符串
   */
  $options = undefined

  /**
   * 获取实例的日期时间片断数据
   *
   * @since 1.0.0
   * @getter
   * @readonly
   * @returns {object}
   */
  get $data() {
    const ctor = this.constructor
    return ctor.getFields(this.$date, this.$format)
  }

  /**
   * 设置实例的日期时间配置项
   *
   * @since 1.0.0
   * @setter
   * @ignore
   */
  set $data(value) {
  }

  /**
   * 获取实例的日期时间配置项
   *
   * @since 1.0.0
   * @getter
   * @readonly
   * @returns {object}
   */
  get $date() {
    return this.$options.date
  }

  /**
   * 设置实例的日期时间配置项
   *
   * @since 1.0.0
   * @setter
   * @ignore
   */
  set $date(value) {
  }

  /**
   * 获取实例的日期时间格式化字符串配置项
   *
   * @since 1.0.0
   * @getter
   * @readonly
   * @returns {string}
   */
  get $format() {
    return this.$options.format
  }

  /**
   * 设置实例的日期时间格式化字符串配置项
   *
   * @since 1.0.0
   * @setter
   * @ignore
   */
  set $format(value) {
  }

  /**
   * 日期时间数据转换成字符串
   *
   * @since 1.0.0
   * @returns {string}
   */
  toString() {
    const ctor = this.constructor
    return ctor.toString(this.$data, this.$format)
  }
}

export default FormatDate

src/timer.js
/* eslint-disable require-jsdoc */
/**
 * 计时器
 *
 * 模块性质：工具模块
 * 作用范围：pc、mobile
 * 依赖模块：utils/format-date，utils/logger
 * 来源项目：扫码点单H5
 * 初始发布日期：2017-05-24 20:00
 * 最后更新日期：2017-05-25 20:00
 *
 * ## 特性
 * - 可检测到时间流速不正常时（如休眠），会自动调整理回正确的时间
 * - 计时器每秒都会可执行一个回调（但不包括最后执行完成时的一个回调，需要通过resolve触发），执行不同的业务逻辑
 * - 返回Promise对象用于操作计时器结束时的回调
 * - 支持递增计时和倒计时（默认）
 * - 时间戳接收值单位为毫秒数，且是1000/单位的整，超出时会向下取整
 * - 会自动抵减时区（比如除了的毫秒数是0，北京时间即08:00:00，会抵减8小时，成00:00:00）
 * - 提供类似码表保存每次触发stop()的时间戳的功能
 *
 * ## Changelog
 * ### 2017.05.29
 * - [feature] 增加递增计时功能
 * - [fix] 修复计时器开始时，回调延迟一秒触发的bug
 *
 * ### 2017.05.28
 * - 某些情况下创建的vue实例未绑定$route属性时的路由名称取值
 *
 * ## Usage
 * ``` js
 * import Timer from 'utils/timer'
 *
 * // 实例化
 * const timer = new Timer({
 *    timeStamp:10*1000
 * })
 *
 * // 倒计时开始
 * timer.start((timerInstance) => {
 *     global.log('每秒执行一次该回调',timerInstance.$datetime)
 * }).then(() => {
 *     global.log('倒计时结束')
 * }).catch(() => {
 *     global.log('中途造成计时中断')
 * })
 *
 * // 倒计时暂停
 * timer.stop().then((timerInstance) => {
 *     global.log('倒计时已暂停')
 * })
 *
 * // 倒计时重置
 * timer.reset().then((timerInstance) => {
 *     global.log('倒计时已重置')
 * })
 * ```
 * @since 3.0.0
 * @version 1.1.0
 */

import formatDate from './format-date'
import Logger from '@~lisfan/logger'

/**
 * 计时器类
 * @class Timer
 */
export default class Timer {
  // 实例配置参数，取默认值
  $options = undefined // 计时器配置选项
  $status = undefined // 计时器当前状态
  $datetime = undefined  // 计时器文本
  $timeouter = undefined // 计时器延迟ID
  $timeZone = undefined // 用户所在时区
  $remainTimeStamp = undefined // 计时器剩余计时时间戳
  $throughTimeStamp = undefined // 计时器已经过的时间戳
  $startTimeStamp = undefined // 计时器开始计时时间戳
  $endTimeStamp = undefined // 计时器开始计时起=>计时器完成时的一轮周期计时时间戳
  $currentTimeStamp = undefined // 当前时间空间维度中正常流逝的时间戳，用于板正流逝的时间
  $stopwatch = [] // 计时器暂停的码表功能
  _logger = undefined // 日志打印器，方便调试

  /**
   * 构造函数
   * @param {object} options - 计时器配置参数
   * @param {number|string} options.timeStamp - 计时时间戳，毫秒旱位，且必须毫秒数为1000单位，不能是1234这样，超出时会自动向下取整
   * @param {boolean} [options.debug] - 是否开启日志调试模式，默认关闭
   * @param {string} [options.format='HH:mm'] - 日期时间格式
   * @param {boolean} [options.isIncrease=false] - 是否为正计时模式，false表示倒计时
   */
  constructor(options) {
    this.$status = 'prepare'
    this.$options = Object.assign({}, Timer.options, options)

    this._logger = new Logger({
      name: 'timer',
      debug: this.$options.debug
    })

    if (!this.$options.timeStamp) {
      this._logger.error('require timeStamp option param, please check!')
    }

    // 当前时间戳
    this.$remainTimeStamp = Math.floor(Number(this.$options.timeStamp) / 1000) * 1000
    // 当前倒计时过去的时间戳
    this.$throughTimeStamp = 0
    // 当前地区的偏移时区
    this.$timeZone = new Date().getTimezoneOffset() * 1000 * 60

    // 当前倒计时间字符串
    // 如果是倒计时，则将当前的时间设置为最大值
    if (this.$options.isIncrease) {
      this.$datetime = formatDate(0, this.$options.format)
    } else {
      this.$datetime = formatDate(this.$remainTimeStamp - this.$timeZone, this.$options.format)
    }
  }

  /**
   * 配置参数
   * @static
   * @enum
   */
  static options = {
    debug: false, // 是否开启调试模式
    timeStamp: undefined,
    format: 'mm:ss',
    isIncrease: false
  }

  /**
   * 更改默认配置参数
   * @static
   *
   * @param {object} [options={}] - 配置对象
   */
  static config(options = {}) {
    Object.assign(Timer.options, options)
  }

  /**
   * 倒计时开始
   * 倒计时结束后进入resolve，倒计时在未结束前中途造成暂停，会触发reject
   * @param {function} callback - 倒计时每秒执行的回调函数
   * @returns {Promise} 返回promise对象
   */
  start(callback) {
    return new Promise((resolve, reject) => {
      this.$status = 'processing'
      // 计时器开始走的这一刻记录当前时间戳
      this.$startTimeStamp = Math.floor(Date.now() / 1000) * 1000
      // 标记倒计时结束的时间戳
      this.$endTimeStamp = this.$startTimeStamp + this.$remainTimeStamp

      const dingdong = () => {
        // 优先执行回调
        if (typeof callback === 'function') {
          callback(this)
        }

        // 如果当前已暂停，则停止倒计时
        if (this.$status === 'stop') {
          return reject(this)
        }

        // 当前的超时时间戳（当currentTimeStamp不在startTimeStamp和endTmeStamp内时，说明已超时）
        // 倒计时也可能在一些途中超时挂起，导致倒计时没有再走下去，所以每次需要重新捕获
        this.$currentTimeStamp = Math.floor(Date.now() / 1000) * 1000

        this._logger.log('currentTimeStamp', new Date(this.$currentTimeStamp).toLocaleString())
        this._logger.log('startTimeStamp', new Date(this.$startTimeStamp).toLocaleString())
        this._logger.log('throughTimeStamp', this.$throughTimeStamp / 1000 + 's')

        if (this.$currentTimeStamp > (this.$startTimeStamp + this.$throughTimeStamp)) {
          this._logger.log('超过时间流速，自动修正!')
          // 超过的时间流速
          const lostTime = this.$currentTimeStamp - this.$startTimeStamp
          this.$remainTimeStamp = this.$options.timeStamp - lostTime - 1000
          this.$throughTimeStamp = lostTime + 1000
        } else {
          this.$remainTimeStamp -= 1000
          this.$throughTimeStamp += 1000
        }

        // 时间已达到最大值，
        if (this.$remainTimeStamp < 0) {
          this.$remainTimeStamp = 0
          this.$throughTimeStamp = this.$endTimeStamp - this.$startTimeStamp
        }

        if (this.$options.isIncrease) {
          /* eslint-disable max-len*/
          this.$datetime = formatDate(this.$options.timeStamp - this.$remainTimeStamp - this.$timeZone, this.$options.format)
        } else {
          const a = this.$remainTimeStamp + this.$timeZone
          console.log('1111', a)
          console.log('222', formatDate(a, this.$options.format))
          this.$datetime = formatDate(this.$remainTimeStamp - this.$timeZone, this.$options.format)
        }
        /* eslint-enable max-len*/

        if (this.$timeouter && this.$currentTimeStamp >= this.$endTimeStamp - 1000) {
          clearTimeout(this.$timeouter)
          this.$timeouter = undefined

          this.$status = 'finished'
          resolve(this)
        } else {
          this.$timeouter = setTimeout(() => {
            dingdong()
          }, 1000)
        }
      }

      dingdong()
    })
  }

  /**
   * 计时器停止
   * @returns {Promise} 返回Promise对象
   */
  stop() {
    clearTimeout(this.$timeouter)
    this.$timeouter = undefined
    this.$status = 'stoped'

    // 记录秒表
    this.$stopwatch.push(this.$datetime)

    return Promise.resolve(this)
  }

  /**
   * 计时器复位
   * @returns {Promise} 返回Promise对象
   */
  reset() {
    clearTimeout(this.$timeouter)
    this.$timeouter = undefined

    this.$status = 'prepare'

    this.$remainTimeStamp = this.$options.timeStamp
    this.$startTimeStamp = Math.floor(Date.now() / 1000) * 1000

    if (this.$options.isIncrease) {
      this.$datetime = formatDate(1, this.$options.format)
    } else {
      this.$datetime = formatDate(this.$remainTimeStamp - this.$timeZone, this.$options.format)
    }

    return Promise.resolve(this)
  }
}
